apollo-entity/src/main/java/com/spotify/apollo/entity/CodecEntityMiddleware.java

```diff
@@ -0,0 +1,250 @@
+  }
+
+  @Override
+  public <R> Middleware<AsyncHandler<R>, AsyncHandler<Response<ByteString>>>
+  asyncSerializerDirect(Class<? extends R> entityClass) {
+    return inner -> inner
+        .map(this::ensureResponse)
+        .map(serialize(entityClass));
+  }
+
+  @Override
+  public <R> Middleware<AsyncHandler<Response<R>>, AsyncHandler<Response<ByteString>>>
+  asyncSerializerResponse(Class<? extends R> entityClass) {
+    return inner -> inner.map(serialize(entityClass));
+  }
+
+  @Override
+  public <E> Middleware<EntityHandler<E, E>, SyncHandler<Response<ByteString>>>
+  direct(Class<? extends E> entityClass) {
+    return direct(entityClass, entityClass);
+  }
+
+  @Override
+  public <E, R> Middleware<EntityHandler<E, R>, SyncHandler<Response<ByteString>>>
+  direct(Class<? extends E> entityClass, Class<? extends R> entityResponseClass) {
+    return inner ->
+        deserialize(entityClass)
+            .flatMap(r -> ctx -> mapPayload(r, inner.invoke(ctx)))
+            .map(serialize(entityResponseClass));
+  }
+
+  @Override
+  public <E> Middleware<EntityResponseHandler<E, E>, SyncHandler<Response<ByteString>>>
+  response(Class<? extends E> entityClass) {
+    return response(entityClass, entityClass);
+  }
+
+  @Override
+  public <E, R> Middleware<EntityResponseHandler<E, R>, SyncHandler<Response<ByteString>>>
+  response(Class<? extends E> entityClass, Class<? extends R> entityResponseClass) {
+    return inner ->
+        deserialize(entityClass)
+            .flatMap(r -> ctx -> flatMapPayload(r, inner.invoke(ctx)))
+            .map(serialize(entityResponseClass));
+  }
+
+  @Override
+  public <E> Middleware<EntityAsyncHandler<E, E>, AsyncHandler<Response<ByteString>>>
+  asyncDirect(Class<? extends E> entityClass) {
+    return asyncDirect(entityClass, entityClass);
+  }
+
+  @Override
+  public <E, R> Middleware<EntityAsyncHandler<E, R>, AsyncHandler<Response<ByteString>>>
+  asyncDirect(Class<? extends E> entityClass, Class<? extends R> entityResponseClass) {
+    return inner ->
+        Middleware.syncToAsync(deserialize(entityClass))
+            .flatMap(r -> ctx -> asyncMapPayload(r, inner.invoke(ctx)))
+            .map(serialize(entityResponseClass));
+  }
+
+  @Override
+  public <E> Middleware<EntityAsyncResponseHandler<E, E>, AsyncHandler<Response<ByteString>>>
+  asyncResponse(Class<? extends E> entityClass) {
+    return asyncResponse(entityClass, entityClass);
+  }
+
+  @Override
+  public <E, R> Middleware<EntityAsyncResponseHandler<E, R>, AsyncHandler<Response<ByteString>>>
+  asyncResponse(Class<? extends E> entityClass, Class<? extends R> entityResponseClass) {
+    return inner ->
+        Middleware.syncToAsync(deserialize(entityClass))
+            .flatMap(r -> ctx -> asyncFlatMapPayload(r, inner.invoke(ctx)))
+            .map(serialize(entityResponseClass));
+  }
+
+  private <E> SyncHandler<Response<E>> deserialize(Class<? extends E> entityClass) {
```

Comment:
The idea is that we can chain the actual processing on to the `Response<E>` created by the deserialization step. If the deserialization fails, this function return a `Response<E>` with no payload and just a error status code with an explanation. Since everything else is operating on the payload, this will short circuit the request and just return the failed deserialization error.